runtime/lua/vim/lsp/buf.lua

```diff
@@ -76,8 +76,76 @@ end
 ---     - on_list: (function) |lsp-on-list-handler| replacing the default handler.
 ---       Called for any non-empty result.
 function M.definition(options)
+  options = options or {}
   local params = util.make_position_params()
-  request_with_options(ms.textDocument_definition, params, options)
+  local bufnr = api.nvim_get_current_buf()
+
+  ---@param results table<integer, {error: lsp.ResponseError, result: lsp.Location|lsp.Location[]|lsp.LocationLink[]|nil}>
+  local function handler(results)
+    local client_id_to_items = {}
+    local total_items = 0
+    for client_id, v in pairs(results) do
+      if v.error == nil and v.result ~= nil then
+        -- first check for Location | LocationLink because
+        -- tbl_islist returns `false` for empty lists
+        if v.result.uri or v.result.targetUri then
+          local client = vim.lsp.get_client_by_id(client_id)
+          local items = util.locations_to_items({ v.result }, client.offset_encoding)
+          total_items = total_items + #items
+          client_id_to_items[client_id] = items
+        elseif vim.tbl_islist(v.result) then
+          local client = vim.lsp.get_client_by_id(client_id)
+          local items = util.locations_to_items(v.result, client.offset_encoding)
+          total_items = total_items + #items
+          client_id_to_items[client_id] = items
+        end
+      end
+    end
+
+    if total_items == 0 then
+      return
+    end
+
+    local items = {}
+
+    -- luacheck: ignore 512
+    for _, v in pairs(client_id_to_items) do
+      for _, v2 in ipairs(v) do
+        table.insert(items, v2)
+      end
+    end
+
+    local title = 'LSP locations'
+    if options.on_list then
+      assert(type(options.on_list) == 'function', 'on_list is not a function')
+      return options.on_list({
+        title = title,
+        items = items,
+        -- FIXME currently passing nil as items can come from multiple clients
+        -- need a better way to attribute items to context
+        context = nil,
```

Comment:
One reason having a corresponding context for each item is to understand which lsp the item is coming from. This can be a part of a custom display of multiple items and doesn't have to be supported in neovim itself. I think this makes sense similar to how you can enable display of diagnostics source.

Reply:
I wonder if providing the context via a separate API (similar to how you can get a client's information given its ID) would be easier/cleaner.

Reply:
> One reason having a corresponding context for each item is to understand which lsp the item is coming from. This can be a part of a custom display of multiple items and doesn't have to be supported in neovim itself. I think this makes sense similar to how you can enable display of diagnostics source.


The current shape/structure won't let you do that because there's no association between a individual item and the context.


We'd either have to change the complete `on_list` parameter shape - which would be breaking, or add a `client_id` to each individual item.

Either way, this doesn't need to block this PR.